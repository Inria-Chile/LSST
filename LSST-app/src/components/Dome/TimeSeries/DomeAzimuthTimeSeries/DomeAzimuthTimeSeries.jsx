import React, { PureComponent } from 'react'
import { LineChart } from 'react-d3-components'
import * as d3 from 'd3';


class DomeAzimuthTimeSeries extends PureComponent {
    static LineColors = {
        'azimuth': '#ddd', 
        'target': '#5e8ba9', 
        'optimal': '#3e6b89'
    }
    
    constructor(props) {
        super(props);
        let defaultData = [...Array(10).keys()].map( (el, i) => {
            return {x:new Date(new Date().getTime() - (10-i)*2000), y: 0};
        });
        this.state = {
            dataPoints: [],
            data: [
                { label: 'target', values: defaultData.slice()}, 
                { label: 'optimal', values: defaultData.slice()},
                { label: 'azimuth', values: defaultData.slice()}, 
            ],
            xScale: d3.scaleTime().domain([new Date(), new Date()]).range([0, this.props.width - 70]),
            yScale: d3.scaleLinear().domain([360, 0]).range([0, this.props.height - 70]),
        };
        this.lineDash = {
            'azimuth': '0', 
            'target': '0', 
            'optimal': '6'
        }
    }

    componentWillReceiveProps() {
        let newData = this.state.data;
        newData[0].values.push({x: new Date(), y: this.props.domeTargetAzimuth})
        newData[1].values.push({x: new Date(), y: this.props.domeOptimalAzimuth})
        newData[2].values.push({x: new Date(), y: this.props.domeAzimuth})
        
        for(let i=0; i<this.state.data.length; ++i){
            if (newData[i].values.length > 10)
                newData[i].values.shift();
        }
        this.setState({
            data: newData,
        })
    }

    render() {
        this.state.xScale.domain([this.state.data[0].values[0].x, new Date()]);
        return (
            <LineChart
                data={this.state.data}
                x={ x => x.x}
                y={ y => y.y}
                width={this.props.width}
                height={this.props.height}
                margin={{top: 10, bottom: 50, left: 50, right: 20}}
                xScale={this.state.xScale}
                yScale={this.state.yScale}
                colorScale={(label) => this.constructor.LineColors[label]}
                stroke={{strokeWidth: (label) => "3", strokeDasharray: (label) => this.lineDash[label]}}
                yAxis={{tickValues: [0,45,90,135,180,225,270,315,360], domain:[0,360]}}
                xAxis={{tickPadding:5, tickArguments: [5], tickFormat: (date) => date.toLocaleTimeString()}}
            />
        );
    }
}


export default DomeAzimuthTimeSeries;