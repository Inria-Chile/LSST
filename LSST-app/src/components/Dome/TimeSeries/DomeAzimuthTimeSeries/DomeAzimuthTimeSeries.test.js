import React from 'react';
import DomeAzimuthTimeSeries from './DomeAzimuthTimeSeries';
import renderer from 'react-test-renderer';
import shallowRenderer from 'react-test-renderer/shallow';

/*---- Snapshot testing ----*/
describe('DomeAzimuthTimeSeries_Test', function(){

  it('renders correctly', () => {
    const renderer = new shallowRenderer();
    const tree = renderer.render(

    <DomeAzimuthTimeSeries domeAzimuth={30}
    domeTargetAzimuth={75}
    domeOptimalAzimuth={40}
    timestamp={new Date()}
    width={600} height={350} />
    );
    expect(tree).toMatchSnapshot();
  });
})

