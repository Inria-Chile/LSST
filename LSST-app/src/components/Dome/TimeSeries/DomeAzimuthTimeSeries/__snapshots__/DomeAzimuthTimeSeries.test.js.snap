// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomeAzimuthTimeSeries_Test renders correctly 1`] = `
<Unknown
  barPadding={0.5}
  colorScale={[Function]}
  data={
    Array [
      Object {
        "label": "target",
        "values": Array [
          Object {
            "x": 2018-01-04T19:01:54.289Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:54.589Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:54.889Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:55.189Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:55.489Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:55.789Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.089Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.389Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.689Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.989Z,
            "y": null,
          },
        ],
      },
      Object {
        "label": "optimal",
        "values": Array [
          Object {
            "x": 2018-01-04T19:01:54.289Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:54.589Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:54.889Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:55.189Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:55.489Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:55.789Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.089Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.389Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.689Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.989Z,
            "y": null,
          },
        ],
      },
      Object {
        "label": "azimuth",
        "values": Array [
          Object {
            "x": 2018-01-04T19:01:54.289Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:54.589Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:54.889Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:55.189Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:55.489Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:55.789Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.089Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.389Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.689Z,
            "y": null,
          },
          Object {
            "x": 2018-01-04T19:01:56.989Z,
            "y": null,
          },
        ],
      },
    ]
  }
  defined={[Function]}
  height={350}
  interpolate="linear"
  label={[Function]}
  margin={
    Object {
      "bottom": 50,
      "left": 50,
      "right": 20,
      "top": 10,
    }
  }
  shape="circle"
  shapeColor={null}
  stroke={
    Object {
      "strokeDasharray": [Function],
      "strokeWidth": [Function],
    }
  }
  tooltipContained={false}
  tooltipHtml={null}
  tooltipMode="mouse"
  tooltipOffset={
    Object {
      "left": 0,
      "top": -35,
    }
  }
  values={[Function]}
  width={600}
  x={[Function]}
  xAxis={
    Object {
      "label": "Time",
      "tickArguments": Array [
        5,
      ],
      "tickFormat": [Function],
      "tickPadding": 5,
    }
  }
  xScale={[Function]}
  y={[Function]}
  y0={[Function]}
  yAxis={
    Object {
      "domain": Array [
        0,
        0,
      ],
      "label": "Angle [deg]",
      "tickArguments": Array [
        7,
      ],
    }
  }
  yScale={[Function]}
/>
`;
